src/services/reader/web_sys.rs

```diff
@@ -0,0 +1,160 @@
+                &reader
+                    .result()
+                    .expect("`FileReader` hasn't finished loading"),
+                0,
+            );
+            let data = FileData {
+                name: name.clone(),
+                content: array.to_vec(),
+            };
+            callback.emit(data);
+        };
+        let listener = Some(EventListener::new(&file_reader, "loadend", callback));
+        file_reader.read_as_array_buffer(&file).unwrap();
+        ReaderTask {
+            file_reader,
+            listener,
+        }
+    }
+
+    /// Reads data chunks from a file and returns them with a callback.
+    pub fn read_file_by_chunks(
+        &mut self,
+        file: File,
+        callback: Callback<FileChunk>,
+        chunk_size: usize,
+    ) -> ReaderTask {
+        let file_reader = FileReader::new().unwrap();
+        let name = file.name();
+        let mut position = 0;
+        let total_size = file.size() as usize;
+        let reader = file_reader.clone();
+        let callback = move |_event: &Event| {
+            let result = reader
+                .result()
+                .expect("`FileReader` hasn't finished loading");
+
+            if result.is_string() {
+                let started = FileChunk::Started { name: name.clone() };
+                callback.emit(started);
+            } else {
+                let array = Uint8Array::new_with_byte_offset(&result, 0);
+                let chunk = FileChunk::DataChunk {
+                    data: array.to_vec(),
+                    progress: position as f32 / total_size as f32,
+                };
+                callback.emit(chunk);
+            };
+            // Read the next chunk
+            if position < total_size {
+                let from = position;
+                let to = cmp::min(position + chunk_size, total_size);
+                position = to;
+                let blob = file.slice_with_i32_and_i32(from as _, to as _).unwrap();
+                reader.read_as_array_buffer(&blob).unwrap();
+            } else {
+                let finished = FileChunk::Finished;
+                callback.emit(finished);
+            }
+        };
+        let listener = Some(EventListener::new(&file_reader, "loadend", callback));
+        let blob = Blob::new().unwrap();
+        file_reader.read_as_text(&blob).unwrap();
+        ReaderTask {
+            file_reader,
+            listener,
+        }
+    }
+}
+
+/// A handle to control reading.
+#[must_use]
+pub struct ReaderTask {
+    file_reader: FileReader,
+    listener: Option<EventListener>,
+}
+
+impl fmt::Debug for ReaderTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("ReaderTask")
+    }
+}
+
+impl Task for ReaderTask {
+    fn is_active(&self) -> bool {
+        self.file_reader.ready_state() == FileReader::LOADING
+    }
+
+    fn cancel(&mut self) {
+        self.file_reader.abort();
```

Comment:
Is there an issue with just checking with `is_active` before we `abort` inside of cancel?
Personally, I would prefer keeping cancel, as long as it can be implemented safely of course.

Reply:
I changed `cancel` to use  `is_active` now before `abort`ing, and removed the check in `drop`, because it calls `cancel`.
Tell me how you like this, or if this solves the problem.

Reply:
Yes, this is what I was hoping for!